src/storage/ynab.ts

```diff
@@ -0,0 +1,103 @@
+import { YNAB_TOKEN, YNAB_BUDGET_ID, YNAB_ACCOUNTS } from "../config.js";
+import { SaveStats, TransactionRow, TransactionStorage } from "../types.js";
+import { createLogger } from "./../utils/logger.js";
+import { parseISO, format } from "date-fns";
+import * as ynab from "ynab";
+import hash from "hash-it";
+
+const YNAB_DATE_FORMAT = "yyyy-MM-dd";
+let ynabAPI: ynab.API;
+
+const logger = createLogger("YNABStorage");
+
+export class YNABStorage implements TransactionStorage {
+  async init() {
+    logger("init");
+  }
+
+  canSave() {
+    return Boolean(YNAB_TOKEN && YNAB_BUDGET_ID);
+  }
+
+  async saveTransactions(txns: Array<TransactionRow>) {
+    await this.init();
+
+    ynabAPI = new ynab.API(YNAB_TOKEN);
+    const budgetName = await getBudgetName(YNAB_BUDGET_ID);
+
+    // Convert transactions to YNAB format
+    logger("transforming transactions to ynab format");
+    const transactionsFromFinancialAccount = txns.map(
+      convertTransactionToYnabFormat,
+    );
+
+    // Send transactions to YNAB
+    logger(`sending to YNAB budget: "${budgetName}"`);
+    await ynabAPI.transactions.createTransactions(YNAB_BUDGET_ID, {
+      transactions: transactionsFromFinancialAccount,
+    });
+    logger("transactions sent to YNAB successfully!");
+
+    const stats: SaveStats = {
+      name: "YNABStorage",
+      table: `budget: "${budgetName}"`,
+      total: txns.length,
+      added: txns.length,
+      pending: NaN,
+      skipped: 0,
+      existing: NaN,
+    };
+
+    return stats;
+  }
+}
+
+function convertTransactionToYnabFormat(
+  tx: TransactionRow,
+): ynab.SaveTransaction {
+  const amount = Math.round(tx.chargedAmount * 1000);
+
+  return {
+    account_id: getYnabAccountIdByAccountNumberFromTransaction(tx.account),
+    date: format(parseISO(tx.date), YNAB_DATE_FORMAT, {}),
+    amount,
+    payee_id: null,
+    payee_name: tx.description,
+    cleared: ynab.TransactionClearedStatus.Cleared,
```

Comment:
You might want to use the `tx.status` to decide if a transaction is cleared or not

Reply:
In YNAB a transaction is marked as "cleared" means that the transaction has been verified or reconciled with your bank or financial institution. So, all the data imported from a bank must be marked as "cleared". 

There is another step into the application when the transactions are "reconciled" with the bank but is optional and is done manually, usually comparing the total balances.

Reply:
We still need to look at `tx.status` as pending transactions won't always have an amount, leading to a different hash.

I guess that in the ynab case we should skip the pending transaction

Reply:
I'll check... YNAB is (in theory) able to import "pending" from banks... but not clear in the API or in the Transaction properties how to indicate that.... I sent an email to support asking for more details.

Reply:
Ok, so YNAB support answered me and confirmed that the API don't support 'pending' transactions. So, maybe we can filter the pending transactions
I actually added filter when there is no correlation between the  tx.account and the uuid in the env vars... so not a lot of works... 
@daniel-hauser wdyt?

Reply:
It's easy to filter out pending transactions - see in https://github.com/daniel-hauser/moneyman/blob/0c91d3549876e3ba3dee2ed0b5fd62a4224bd47a/src/storage/sheets.ts#L102C7-L106C8

Reply:
I modified the filter and the way to calculate the statistics to avoid all the extra maps. In addition, added a warning in case there is a mismatch between the accountId and the Ynab account ID.